lms/djangoapps/course_blocks/transformers/library_content.py

```diff
@@ -177,3 +188,82 @@ def publish_event(event_name, result, **kwargs):
             format_block_keys,
             publish_event,
         )
+
+
+class ContentLibraryOrderTransformer(BlockStructureTransformer):
+    """
+    A transformer that manipulates the block structure by modifying the order of the
+    selected blocks within a library_content module to match the order of the selections
+    made by the ContentLibraryTransformer or the corresponding XBlock. So this transformer
+    requires the selections for the randomized content block to be already
+    made either by the ContentLibraryTransformer or the XBlock.
+
+    Staff users are *not* exempted from library content pathways/
+    """
+    WRITE_VERSION = 1
+    READ_VERSION = 1
+
+    @classmethod
+    def name(cls):
+        """
+        Unique identifier for the transformer's class;
+        same identifier used in setup.py
+        """
+        return "library_content_randomize"
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Collects any information that's necessary to execute this
+        transformer's transform method.
+        """
+        block_structure.request_xblock_fields('mode')
+        block_structure.request_xblock_fields('max_count')
+
+        ContentLibraryTransformer.set_block_analytics_summary(block_structure, modulestore())
```

Comment:
What happens in the initial deployment phase, when the `collect()` here has not yet run for this course? I forget if we lazily re-load those or we need to do something special (I've got some docs, but I'm not sure if they're up to date).

Reply:
@ormsbee, the code in this `collect()` method is exactly the same as the code in the `collect()` method of the `ContentLibraryTransformer` class. Just that the code in the latter which sets the `block_analytics_summary` transformer block field [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_blocks/transformers/library_content.py#L53-L72) has been refactored into a separate class method so that it can be reused by the `collect()` method of the `ContentLibraryOrderTransformer` class.

> What happens in the initial deployment phase, when the collect() here has not yet run for this course? I forget if we lazily re-load those or we need to do something special (I've got some docs, but I'm not sure if they're up to date).

So to answer your question, the `set_block_analytics_summary()` code will not be executed unless the `collect()` method for this transformer or the `ContentLibraryTransformer` is executed.

My understanding of the block structure is extremely limited, so please let me know if I am missing something here. This method contains the same code as the existing code in the `ContentLibraryTransformer.collect()` method (just refactored to allow reuse) and does conform to the guideline mentioned in the `BlockStructureTransformer` base class [here](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/block_structure/transformer.py#L77-L107).

The only dependency this transformer has is on the `LibraryContentTransformer` to do the selection before the order transformation can be done. I have enforced this dependency by the ordering [here](https://github.com/open-craft/edx-platform/blob/guruprasad/BB-112-fix-randomized-content-block-order/lms/djangoapps/course_blocks/api.py#L41-L42) and in `setup.py` [here](https://github.com/open-craft/edx-platform/blob/guruprasad/BB-112-fix-randomized-content-block-order/setup.py#L53-L54). The `collect()` methods of the transformers are called [here](https://github.com/open-craft/edx-platform/blob/guruprasad/BB-112-fix-randomized-content-block-order/openedx/core/djangoapps/content/block_structure/transformers.py#L79-L81) in the order returned by the `stevedore` `ExtensionManager`. However, I am not sure how the order returned by that will be and if that is repeatable. I tried digging into `stevedore` and into `setuptools`, `pkg_resources` etc., but couldn't figure it out.

The order returned in our devstacks and a few production environments seem to load the `ContentLibraryTransformer` before the `ContentLibraryOrderTransformer` and hence we have not seen any issues at all.

However, I am not familiar with the setup in edx.org or the course(s) which started throwing errors after this was merged twice or have access to logs to be able to debug this further.

> I've got some docs, but I'm not sure if they're up to date).

You're the subject-matter expert here, so I will accept what you find and recommend :)

Reply:
@ormsbee, did you get a chance to check my replies?